resource-managers/kubernetes/token-refresh-server/src/main/scala/org/apache/spark/security/kubernetes/TokenRefreshService.scala

```diff
@@ -0,0 +1,349 @@
+        maxDate <= nowMills) {
+        logDebug(s"Obtaining a new token with maxData $maxDate," +
+          s" expireTime $expireTime, now $nowMills")
+        val newToken = obtainNewToken(token, identifier)
+        logInfo(s"Obtained token $newToken")
+        newToken
+      } else {
+        token
+      }
+    } else {
+      token
+    }
+    maybeNewToken
+  }
+
+  private def maybeRenewExpireTime(token: Token[_ <: TokenIdentifier], expireTime: Long,
+                                   deadline: Long, nowMillis: Long) = {
+    if (expireTime <= deadline || expireTime <= nowMillis) {
+      try {
+        logDebug(s"Renewing token $token with current expire time $expireTime," +
+          s" deadline $deadline, now $nowMillis")
+        val newExpireTime = token.renew(hadoopConf)
+        logDebug(s"Renewed token $token. Next expire time $newExpireTime")
+        newExpireTime
+      } catch {
+        case t: Throwable =>
+          logWarning(t.getMessage, t)
+          hasError = true
+
+          getRetryTime
+      }
+    } else {
+      expireTime
+    }
+  }
+
+  private def obtainNewToken(token: Token[_ <: TokenIdentifier],
+                             identifier: AbstractDelegationTokenIdentifier) = {
+    val owner = identifier.getOwner
+    val realUser = identifier.getRealUser
+    val user = if (realUser == null || realUser.toString.isEmpty || realUser.equals(owner)) {
+      owner.toString
+    } else {
+      realUser.toString
+    }
+    val credentials = new Credentials()
+    val ugi = UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser)
+    val newToken = ugi.doAs(new PrivilegedExceptionAction[Token[_ <: TokenIdentifier]] {
+
+      override def run() : Token[_ <: TokenIdentifier] = {
+        val fs = FileSystem.get(hadoopConf)
+        val tokens = fs.addDelegationTokens(UserGroupInformation.getLoginUser.getUserName,
+          credentials)
+        tokens(0)
+      }
+    })
+    newToken
+  }
+
+  private def writeTokensToSecret(tokenToExpire: Map[Token[_ <: TokenIdentifier], Long],
+                                  nowMillis: Long) : Unit = {
+    val durationUntilExpire = tokenToExpire.values.min - nowMillis
+    val key = s"$SECRET_DATA_ITEM_KEY_PREFIX_HADOOP_TOKENS$nowMillis-$durationUntilExpire"
+    val credentials = new Credentials()
+    tokenToExpire.keys.foreach(token => credentials.addToken(token.getService, token))
+    val serialized = serializeCredentials(credentials)
+    val value = Base64.encodeBase64String(serialized)
+    val secretMeta = renew.secretMeta
+    val editor = kubernetesClient.secrets
+      .inNamespace(secretMeta.getNamespace)
+      .withName(secretMeta.getName)
+      .edit()
+    editor.addToData(key, value)
+    val dataItemKeys = editor.getData.keySet().asScala.filter(
+      _.startsWith(SECRET_DATA_ITEM_KEY_PREFIX_HADOOP_TOKENS)).toSeq.sorted
+    // Remove data items except the latest two data items.
```

Comment:
A k8s secret can hold only up to 1 MB data. So we need to remove old data items. I kept two items here to avoid race conditions where some newly launching executors may access the previous token.